### PR TITLE
Unsubscribe reducer

### LINKED ISSUE #7853: Add reducer code/saga code for unsubscribing a user from a notification
Supports #7656

---

Implement the logic to actually perform the unsubscribe action when the page loads.

### PR DESCRIPTION
**Pull request by willdurand**
Fixes #7853
~~⚠️ depends on #7854 and #7844~~

---

This patch adds the reducer/saga logic for the unsubscribe page.

### CODE REVIEW

File: src/amo/reducers/users.js [hunk 1]
```diff
@@ -475,15 +518,99 @@ export const addUserToState = ({
...
+  ];
+};
+
 type Action =
+  | AbortUnsubscribeNotificationAction
   | FetchUserAccountAction
   | FetchUserNotificationsAction
   | FinishUpdateUserAccountAction
-  | UpdateUserAccountAction
+  | FinishUpdateUserAccountAction
```
Comment (bobsilverberg) [inline src/amo/reducers/users.js]: This appears to be repeated, whereas `FinishUnsubscribeNotificationAction` does not appear in this list. I wonder why this is not generating any errors?

**Comment by willdurand** [comment]
This is now rebased, you can look at the UI implementation in #7857.

File: tests/unit/amo/reducers/test_users.js [hunk 2]
```diff
@@ -445,4 +450,172 @@ describe(__filename, () => {
...
+
+      expect(isUnsubscribedFor(state, hash, notification, token)).toEqual(
+        false,
+      );
+    });
+
+    it('does not change the value of another notification', () => {
+      const hash = 'some-hash';
+      const token = 'some-token';
+      const notification1 = 'notification1';
```
Comment (bobsilverberg) [inline tests/unit/amo/reducers/test_users.js]: To improve readability of the test, you might consider renaming `notification1` to `updatedNotification` and `notification2` to `pendingNotification`.

File: tests/unit/amo/reducers/test_users.js [hunk 3]
```diff
@@ -445,4 +450,172 @@ describe(__filename, () => {
...
+        }),
+      );
+
+      expect(isUnsubscribedFor(state, hash, notification, token)).toEqual(true);
+    });
+
+    it('does not change the value of another notification', () => {
+      const hash = 'some-hash';
+      const token = 'some-token';
+      const notification1 = 'notification1';
```
Comment (bobsilverberg) [inline tests/unit/amo/reducers/test_users.js]: As above, to improve readability of the test, you might consider renaming `notification1` to `pendingNotification` and `notification2` to `updatedNotification`.

File: tests/unit/amo/reducers/test_users.js [hunk 4]
```diff
@@ -445,4 +450,172 @@ describe(__filename, () => {
...
+        false,
+      );
+      expect(isUnsubscribedFor(state, hash, notification2, token)).toEqual(
+        true,
+      );
+    });
+  });
+
+  describe('abortUnsubscribeNotification', () => {
+    it('sets `isUnsubscribedFor` to `true`', () => {
```
Comment (bobsilverberg) [inline tests/unit/amo/reducers/test_users.js]: I think this should read "sets `isUnsubscribedFor` to `null`".

File: tests/unit/amo/reducers/test_users.js [hunk 5]
```diff
@@ -445,4 +450,172 @@ describe(__filename, () => {
...
+        }),
+      );
+
+      expect(isUnsubscribedFor(state, hash, notification, token)).toEqual(null);
+    });
+
+    it('does not change the value of another notification', () => {
+      const hash = 'some-hash';
+      const token = 'some-token';
+      const notification1 = 'notification1';
```
Comment (bobsilverberg) [inline tests/unit/amo/reducers/test_users.js]: As above, to improve readability of the test, you might consider renaming `notification1` to `pendingNotification` and `notification2` to `abortedNotification`.

File: src/amo/reducers/users.js (hunk 1 above)
Comment (willdurand) [reply src/amo/reducers/users.js]: oh oops. I think we need a special Flow notation to get an error in the reducer.

File: tests/unit/amo/reducers/test_users.js (hunk 3 above)
Comment (willdurand) [reply tests/unit/amo/reducers/test_users.js]: oh, great idea! 🥇

**Pull request merged**